Add tutorial about @algobuilder/web (#649)
scale-it/algo-builder: docs/tutorials/t-08.md

## Objective

This tutorial is to demonstrate [`Algosigner`](https://github.com/PureStake/algosigner) and [`Wallet Connect`](https://walletconnect.com/) in webapp using [`@algo-builder/web`](https://github.com/scale-it/algo-builder/tree/master/packages/web). [`@algo-builder/web`](https://github.com/scale-it/algo-builder/tree/master/packages/web) is designed to be used with web dapps as well as scripts and user programs and also allows you to interact with contracts easily.

## What is `@algo-builder/web`?

> This package provides a class `WebMode` which has variety of high level functions like, [`waitForConfirmation`](algobuilder.dev/api/web/classes/web.html#waitForConfirmation), [`executeTx`](<(algobuilder.dev/api/web/classes/web.html#executeTx)>), [`signTransaction`](<(algobuilder.dev/api/web/classes/web.html#signTransaction)>), etc. These functions help sending transactions and building dapps.

Reviewer: I can't asset to link `waitForConfirmation`, `executeTx` and `signTransaction`. Could you check it

Author: Have updated the link

## Requirements

- Node 14+
- Clone [`wallet-connect-template`]().[TODO: Add template URL when template PR is merged.]. This is a Vue3 app.

> - `pipenv` with our `Pipfile`. Read more about [`pipenv`](#Using-Pipenv).

Reviewer: the link to use `pipenv` is wrong.

Author: link updated

> - Yarn `v3.1+` or NPM `v8.0+` or PNPM `v6.21+`.

Reviewer: Not sure about yarn version.

Author: I have tried with 2 and it does not work.

- Setup the accounts on [`Algosigner`](https://github.com/PureStake/algosigner) and [`Wallet Connect`](https://walletconnect.com/) wallet. It could be MainNet, TestNet or private net.

## Steps

> 1. Setup the [`wallet-connect-template`](). [TODO: Add template URL when template PR is merged.]

Reviewer: Let comment TODO

Author: I have added the link which will work once template PR is merged.

> 2. The above template should be accessbile on [`localhost`](http://localhost:8080/) and you shoulb be able to see below screen with dropdown.

Reviewer: ```suggestion
2. The above template should be accessbile on your localhost with port 8080 and you should be able to see below screen with drop down.
```

Reviewer: Or you can do with better way as you refer ❤️

Author: I would prefer to link the localhost URL so that the user can directly access it without having to enter into the browser. I have committed the other suggestion apart from this.

<img src="./t-08/assets/select-wallet.png" height="200" title="Select Wallet" />

3. Click on the dropdown and you should be able to see below screen. Select any wallet you prefer. 

<img src="./t-08/assets/all-wallets.png" height="200" title="All Wallet" />

>     Below is the `switch case` which gets triggered on wallet change. This tries to connect with wallet and display the address.

Reviewer: The file reader like below on my machine.

![image](https://user-images.githubusercontent.com/45826131/162144471-035a80f5-7368-4f81-8946-b5fc8b970a34.png)

Are you have same problem?

Author: Yes, on VS code it renders perfectly fine but on Github it is misalgined. I have fixed it now. Please check.

Reviewer: Looks good now

>    Below is the `switch case` which gets triggered on wallet change. This tries to connect with wallet and displays the address.

Reviewer 2: ```suggestion
   Below is a `switch` statement which gets triggered on wallet change. It tries to connect with wallet and displays the address.
```

```js
   switch (e.target.value) {
     case WalletType.ALGOSIGNER: // AlgoSigner
       this.connectAlgoSigner();
       break;
     case WalletType.MY_ALGO: // My Algo Wallet
       this.connectMyAlgoWallet();
       break;
     case WalletType.WALLET_CONNECT: // Wallet Connect
       this.connectWalletConnect();
       break;
     default:
       console.warn("Wallet %s not supported", e.target.value);
   }
```

>    Below method `connectAlgoSigner()` is used to connect the AlgoSigner wallet. To use `web` package in your Vue app, first you need to create an instance of the `WebMode` class by passing `AlgoSigner` and the chain name.
> 
> ### Example
> 

Reviewer 2: ```suggestion
### Example

We use `connectAlgoSigner()` method to connect the AlgoSigner wallet. Next, we create an instance of the `WebMode` class by passing and pass `AlgoSigner` and the chain name. The `web` instance will allow us to send and sign transactions to the blockchain.
```

 ```js
   const web = new WebMode(AlgoSigner, CHAIN_NAME);
 ```

```js
   async connectAlgoSigner() {
      try {
        const webMode = new WebMode(AlgoSigner, ChainType.MainNet);
        console.log("WebMode initialized: ", webMode);
        const algoSignerResponse = await AlgoSigner.connect({
          ledger: ChainType.MainNet,
        });
        this.setWalletType(WalletType.ALGOSIGNER);
        console.log("Connected to AlgoSigner:", algoSignerResponse);
        await this.getUserAccount();
      } catch (e) {
        console.error(e);
      }
   }
```

>    Below method `connectMyAlgoWallet()` is used to connect the My Algo wallet.

Reviewer 2: ```suggestion
`connectMyAlgoWallet()` method is used to connect the My Algo wallet.
```

```js
   async connectMyAlgoWallet() {
      try {
        let myAlgo = new MyAlgoWalletSession(ChainType.MainNet);
        await myAlgo.connectToMyAlgo();
        if (myAlgo.accounts.length) {
          this.walletAddress = myAlgo.accounts[0].address;
        }
      } catch (e) {
        console.error(e);
      }
   }
```

>    Below method `connectWalletConnectr()` is used to connect the Wallet Connect wallet.

Reviewer: ```suggestion
   Below method `connectWalletConnect()` is used to connect the Wallet Connect wallet.
```

```js
   async connectWalletConnect() {
      try {
        let walletConnector = new WallectConnectSession(ChainType.MainNet);
        await walletConnector.create(true);
        walletConnector.onConnect((error, response) => {
          if (response.accounts.length) {
            this.walletAddress = response.accounts[0];
          }
        });
      } catch (e) {
        console.error(e);
      }
   }
```

> 4. Once you have initialzed the `WebMode` instance. Now you can use it to execute a transaction:

Reviewer 2: ```suggestion
4. Once you have initialzed the `WebMode` instance, you can use it to execute a transaction:
```

    ```js
    const txParams = {
      type: types.TransactionType.TransferAlgo,
      sign: types.SignType.SecretKey,
      fromAccountAddr: fromAddress,
      toAccountAddr: toAddress,
      amountMicroAlgos: amount,
      payFlags: {},
    };
    let response = await web.executeTx(txParams);
    ```

    The above code will make the transaction, let the user sign it using algosigner and send it to the network. You can also use `web.sendTransaction()` or `web.signTransaction()` in a vue app.

> 5. Once your selected wallet is connect. Now you can see you address like below on successful connection.

Reviewer 2: ```suggestion
5. Once your selected wallet is connect., you can see you address like below on successful connection.
```
